### PR TITLE
[ELF] Add --compress-section to compress matched non-SHF_ALLOC sections

### PR DESCRIPTION
**Pull request by MaskRay**
--compress-sections <section-glib>=[none|zlib|zstd] is similar to
--compress-debug-sections but applies to broader sections without the
SHF_ALLOC flag. lld will report an error if a SHF_ALLOC section is
matched. An interesting use case is to compress `.strtab`/`.symtab`,
which consume a significant portion of the file size (15.1% for a
release build of Clang).

An older revision is available at https://reviews.llvm.org/D154641 .
This patch focuses on non-allocated sections for safety. Moving
`maybeCompress` as D154641 does not handle STT_SECTION symbols for
`-r --compress-debug-sections=zlib` (see `relocatable-section-symbol.s`
from #66804).

Since different output sections may use different compression
algorithms, we need CompressedData::type to generalize
config->compressDebugSections.

GNU ld feature request: https://sourceware.org/bugzilla/show_bug.cgi?id=27452

Link: https://discourse.llvm.org/t/rfc-compress-arbitrary-sections-with-ld-lld-compress-sections/71674

### CODE REVIEW

File: lld/ELF/OutputSections.cpp [hunk 1]
```diff
@@ -450,20 +470,18 @@ void OutputSection::writeTo(uint8_t *buf, parallel::TaskGroup &tg) {
   if (type == SHT_NOBITS)
     return;
 
-  // If --compress-debug-section is specified and if this is a debug section,
-  // we've already compressed section contents. If that's the case,
-  // just write it down.
+  // If the section is compressed due to
+  // --compress-debug-section/--compress-sections, the content is already known.
```
Comment (petrhosek) [inline lld/ELF/OutputSections.cpp:474]:
```suggestion
  // --compress-debug-section/--compress-nonalloc-sections, the content is already known.
```

**Comment by petrhosek** [comment]
I'd prefer `--compress-sections` as a flag name and mention in the help and documentation that it's only applicable to non-allocated sections. That gives us an option to eventually extend the implementation to allocated sections without having to introduce yet another flag (plus `--compress-sections` is easier to spell and remember).

**Comment by igorkudrin** [comment]
It seems a bit inconsistent to me that, say, `--compress-nonalloc-sections '*bug*=zstd'` can be overridden by `--compress-debug-sections=zlib`, but not by `--compress-debug-sections=none`.

**Comment by MaskRay** [comment]
> It seems a bit inconsistent to me that, say, `--compress-nonalloc-sections '*bug*=zstd'` can be overridden by `--compress-debug-sections=zlib`, but not by `--compress-debug-sections=none`.

`=none` can override a previous `=zlib` or `=zstd`.
`ld.lld -pie %t.o -o %t --compress-nonalloc-sections '*0=zlib' --compress-nonalloc-sections '*0=none'` in `lld/test/ELF/compress-nonalloc-sections.s` is such an example.

**Comment by igorkudrin** [comment]
> > It seems a bit inconsistent to me that, say, `--compress-nonalloc-sections '*bug*=zstd'` can be overridden by `--compress-debug-sections=zlib`, but not by `--compress-debug-sections=none`.
> 
> `=none` can override a previous `=zlib` or `=zstd`. `ld.lld -pie %t.o -o %t --compress-nonalloc-sections '*0=zlib' --compress-nonalloc-sections '*0=none'` in `lld/test/ELF/compress-nonalloc-sections.s` is such an example.

And still, if the name template in the `--compress-nonalloc-sections` matches debug sections, you cannot use just `--compress-debug-sections=none` to switch compressing debug sections off. Yes, there are possible workarounds to do that, but the most straightforward variant doesn't work.

**Comment by MaskRay** [comment]
> > > It seems a bit inconsistent to me that, say, `--compress-nonalloc-sections '*bug*=zstd'` can be overridden by `--compress-debug-sections=zlib`, but not by `--compress-debug-sections=none`.
> > 
> > 
> > `=none` can override a previous `=zlib` or `=zstd`. `ld.lld -pie %t.o -o %t --compress-nonalloc-sections '*0=zlib' --compress-nonalloc-sections '*0=none'` in `lld/test/ELF/compress-nonalloc-sections.s` is such an example.
> 
> And still, if the name template in the `--compress-nonalloc-sections` matches debug sections, you cannot use just `--compress-debug-sections=none` to switch compressing debug sections off. Yes, there are possible workarounds to do that, but the most straightforward variant doesn't work.

Sorry, I was confused. I think you are talking about the option precedence. I agree that --compress-debug-sections should win, even if it is `none`.
Added a test that `--compress-debug-sections=none --compress-sections .debug_str=zstd` that disables compression for `.debug_str`.

**Review by igorkudrin — APPROVED** [review]
LGTM, thanks

**Comment by MaskRay** [comment]
This patch has been reviewed last July and we have decided to have at least the non-SHF_ALLOC form.
I somehow tried to figure out the SHF_ALLOC issue and did not land it. I'll land this specific non-SHF_ALLOC form shortly.

**Comment by frobtech** [comment]
I think the only concern about the "warn for SHF_ALLOC match" behavior is if someone wanted to use a catch-all (or otherwise very wide) pattern with the expectation of it omitting SHF_ALLOC sections whose names match. But I think that's a new outlying use that we can address in later iterations. The warning behavior seems like a conservative and safe way to add the feature initially.

**Pull request merged**